core/src/transport/incoming_limit.rs

```diff
@@ -0,0 +1,401 @@
+    }
+}
+
+#[must_use]
+#[pin_project::pin_project(PinnedDrop)]
+pub struct IncomingLimitListen<TInner> {
+    #[pin]
+    inner: TInner,
+    limit: Arc<IncomingLimit>,
+    /// Entry number reserved to us within `limit.inner.wakers`. Starts with `None`, then
+    /// assigned only if necessary. Once it is `Some`, the value is never changed.
+    waker_slot_num: Option<usize>,
+}
+
+impl<TInner, TUpgr> Stream for IncomingLimitListen<TInner>
+where
+    TInner: TryStream<Ok = ListenerEvent<TUpgr>>,
+{
+    type Item = Result<ListenerEvent<IncomingLimitUpgrade<TUpgr>>, TInner::Error>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+        let this = self.project();
+
+        // We increase `current_num_approx` by one.
+        // The `_guard` will make sure that we decrease it again before leaving this function.
+        let (_guard, old_current_num_approx) = NumApproxGuard::acquire(&this.limit);
+        if old_current_num_approx >= this.limit.limit {
+            // `current_num_approx` is over the limit. We lock the mutex in order to check
+            // whether we are actually indeed over the limit.
+            let mut inner = this.limit.inner.lock();
+            if inner.current_num >= this.limit.limit {
+                // We are over the limit. Register a waker and return `Pending`.
+                if let Some(waker_slot_num) = this.waker_slot_num {
+                    let our_slot = &mut inner.wakers[*waker_slot_num];
+                    if our_slot.as_ref().map(|s| !cx.waker().will_wake(s)).unwrap_or(true) {
+                        *our_slot = Some(cx.waker().clone());
+                    }
+                } else {
+                    let waker_slot_num = inner.wakers.insert(Some(cx.waker().clone()));
+                    *this.waker_slot_num = Some(waker_slot_num);
+                }
+
+                return Poll::Pending;
+            }
+        };
+
+        // We know that we are under the limit. We can poll the listener.
+        // Note that `_guard` is still alive here.
+        let (upgrade, local_addr, remote_addr) = match this.inner.try_poll_next(cx) {
+            Poll::Pending => return Poll::Pending,
+            Poll::Ready(None) => return Poll::Ready(None),
+            Poll::Ready(Some(Ok(ListenerEvent::NewAddress(addr)))) =>
+                return Poll::Ready(Some(Ok(ListenerEvent::NewAddress(addr)))),
+            Poll::Ready(Some(Ok(ListenerEvent::AddressExpired(addr)))) =>
+                return Poll::Ready(Some(Ok(ListenerEvent::AddressExpired(addr)))),
+            Poll::Ready(Some(Err(err))) =>
+                return Poll::Ready(Some(Err(err))),
+            Poll::Ready(Some(Ok(ListenerEvent::Upgrade { upgrade, local_addr, remote_addr }))) => {
+                (upgrade, local_addr, remote_addr)
+            }
+        };
+
+        // Properly reserve our slot. Requires a mutex lock.
+        mem::forget(_guard);
```

Comment:
Should we add a comment stating that at this point we do not want the destructor to run to not decrement the counter since we now have a new incoming connection?